### PR TITLE
[WIP] updated build setup and actors chat, client and dir

### PR DESCRIPTION
**Pull request by hgoldhammer**
updated BehaviorFactory

### CODE REVIEW

File: akka/src/main/scala/chatapp/Chat.scala [hunk 1]
```diff
@@ -16,81 +16,79 @@ object Action extends Enumeration {
 import Action._
 
 class BehaviorFactory(compute: Int, post: Int, leave: Int, invite: Int) {
-  def apply(dice: DiceRoll): Action = {
-    val pick = dice()
+  val _compute: Int = compute
+  val _post: Int = _compute + post
+  val _leave: Int = _post + leave
+  val _invite: Int = _leave + invite
```
Comment (smarr) [inline akka/src/main/scala/chatapp/Chat.scala]: Why is this needed?
The arguments to the constructor are visible in the body of the class for direct use, no?

File: akka/src/main/scala/chatapp/Chat.scala [hunk 2]
```diff
@@ -16,81 +16,79 @@ object Action extends Enumeration {
...
+
+  def apply(pick: Int): Action = {
```
Comment (smarr) [inline akka/src/main/scala/chatapp/Chat.scala]: Why this change? Pony passes also the `DiceRoll` object to the apply method: https://github.com/sblessing/chat-app/blob/master/pony/chat.pony#L47

File: akka/src/main/scala/chatapp/Chat.scala [hunk 3]
```diff
@@ -16,81 +16,79 @@ object Action extends Enumeration {
...
     action
   }
 }
 
 
 sealed trait ChatMsg
-final case class ChatPost(payload: Array[Byte], accumulator: ActorRef[Accumulator]) extends ChatMsg
-final case class ChatJoin(client: ActorRef[Client], accumulator: ActorRef[Accumulator]) extends ChatMsg
-final case class ChatLeave(client: ActorRef[Client], didLogout: Boolean, accumulator: ActorRef[Accumulator]) extends ChatMsg
+final case class ChatPost(payload: Array[Byte], accumulator: ActorRef[AccMsg]) extends ChatMsg
```
Comment (smarr) [inline akka/src/main/scala/chatapp/Chat.scala]: Hm, am I misunderstanding what an ActorRef's type should be?
It's the message type not the type of the actor object?

Is there any way for me to get this understanding from the docs?
Looking at https://doc.akka.io/api/akka/current/akka/actor/typed/ActorRef.html
T is not explicitly documented.

Though, I see it used in some method signatures to type messages.
So, I guess your change is correct. Still strange to me.

File: akka/src/main/scala/chatapp/Chat.scala [hunk 4]
```diff
@@ -99,123 +97,125 @@ class Chat(initiator: ActorRef[Client]) {
...
 
-class Client(id: Long, directory: Directory, seed: Long) {
+class Client(id: Long, directory: ActorRef[DirMsg], seed: Long) {
 
-  val friends = mutable.ArrayBuffer[ActorRef[Client]]()
-  val chats   = mutable.ArrayBuffer[ActorRef[Chat]]()
+  val friends = new mutable.ArrayBuffer[ActorRef[ClientMsg]]()
+  val chats   = new mutable.ArrayBuffer[ActorRef[ChatMsg]]()
   val rand = new SimpleRand(seed)
-  val dice = new DiceRoll(rand)
```
Comment (smarr) [inline akka/src/main/scala/chatapp/Chat.scala:121]: Why removing the `DiceRoll`?
Pony also still has it: https://github.com/sblessing/chat-app/blob/master/pony/chat.pony#L122
Would be good to make things consistent, and avoid having too many variations of things.

File: akka/src/main/scala/chatapp/Chat.scala [hunk 5]
```diff
@@ -99,123 +97,125 @@ class Chat(initiator: ActorRef[Client]) {
...
   var fibIndex = 35
 
-  def apply(): Behavior[ClientMsg] =
-    Behaviors.setup { context =>
-      Behaviors.receiveMessage {
```
Comment (smarr) [inline akka/src/main/scala/chatapp/Chat.scala:126]: Why the change here?
Behaviors.setup + Behaviors.receiveMessage where used in the example for Akka typed.

File: akka/src/main/scala/chatapp/Chat.scala [hunk 6]
```diff
@@ -240,48 +240,96 @@ class Client(id: Long, directory: Directory, seed: Long) {
...
+final case class DirLogin(id: Long) extends DirMsg
+final case class DirBefriend() extends DirMsg
+final case class DirLeft(client: ActorRef[ClientMsg]) extends DirMsg
+final case class DirPoke(factory: BehaviorFactory, accumulator: ActorRef[AccMsg]) extends DirMsg
+final case class DirDisconnect(poker: ActorRef[PokerMsg]) extends DirMsg
 
 class Directory(seed: Long, befriend: Int) {
-  val clients = mutable.ArrayBuffer[ActorRef[Client]]()
-  val random = SimpleRand(seed)
-  var poker: ActorRef[Poker] = null
```
Comment (smarr) [inline akka/src/main/scala/chatapp/Chat.scala:252]: What's wrong with the name `poker`? And what's the `_` instead of the `null` here?

**Review by smarr — COMMENTED** [review]
Most of the comments are simply my missing knowledge around Akka.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 1 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: see https://github.com/sblessing/chat-app/blob/ce80afdfea43cad366bb8b0ccd6341be9da22fa8/pony/chat.pony#L42
Yes, the arguments are visible in the body.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 2 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: In CAF we changed it that way to have only one random generator, pony use only one too.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 3 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: I used the reference docs, see https://doc.akka.io/docs/akka/current/typed/actors.html
You use `sealed trait` to group message for one actor type, thats defines the `Behavior`  and `ActorRef` type.
You can have more then one `object` or `class` that defines behavior for one `sealed trait T`

File: akka/src/main/scala/chatapp/Chat.scala (hunk 4 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala:121]: see https://github.com/smarr/chat-app/pull/1#discussion_r458927685

File: akka/src/main/scala/chatapp/Chat.scala (hunk 6 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala:252]: In the variable scop the name poker is used twice. `_` and `null` can often used both, I changed it because my IDE marked it.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 5 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala:126]: When you only receive messages than I use `Behavoirs.receive`, when you want to first calculation or spawn new actor than first the setup and than start receiving. The examples are not constant to that point.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 1 above)
Comment (smarr) [reply akka/src/main/scala/chatapp/Chat.scala]: I don't understand this reply.
In Pony, you got a constructor and need to initialize the fields explicitly.
In Scala, this is not necessary, because the constructor's arguments are always accessible. They are not in Pony.

Introducing these fields seems unnecessary.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 2 above)
Comment (smarr) [reply akka/src/main/scala/chatapp/Chat.scala]: Ehm, here we only got one random number generator, too.
`DiceRoll` is simply a decorator for the PRNG. As it is in Pony.
To stay in sync with Pony, I think this change is also better reverted.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 6 above)
Comment (smarr) [reply akka/src/main/scala/chatapp/Chat.scala:252]: Which IDE are you using? I am currently having the Metals plugin for VS code, to have Pony and the rest all in one thing.

Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala:252]: I us IntelliJ IDEA with scala plugin, only for akka

File: akka/src/main/scala/chatapp/Chat.scala (hunk 1 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: I added the fields because of the calculations to set them to same value as pony and CAF

File: akka/src/main/scala/chatapp/Chat.scala (hunk 2 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: I used CAF as reference for the rework, I thought that was changed to pony too.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 1 above)
Comment (smarr) [reply akka/src/main/scala/chatapp/Chat.scala]: Ah, my bad. That's my sloppy reading of the code. I only saw the initialization.

File: akka/src/main/scala/chatapp/Chat.scala (hunk 2 above)
Comment (smarr) [reply akka/src/main/scala/chatapp/Chat.scala]: I think we refer to Pony as the reference implementation, and please see my link above, it's still there, no?

Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: ok, I will change it back.

Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala]: done

File: akka/src/main/scala/chatapp/Chat.scala (hunk 4 above)
Comment (hgoldhammer) [reply akka/src/main/scala/chatapp/Chat.scala:121]: done

Comment (smarr) [reply akka/src/main/scala/chatapp/Chat.scala:121]: thanks

**Review by smarr — APPROVED** [review]
Thanks for adding the formatting!
